python/packages/zarathustra/skills/subgraph_query_abci/rounds.py

```diff
@@ -57,31 +61,56 @@ class SynchronizedData(BaseSynchronizedData):
     This data is replicated by the tendermint application.
     """
 
+    @property
+    def most_voted_subgraph_config(self):
+        return self.db.get_strict("most_voted_subgraph_config")
+
+    @property
+    def most_voted_subgraph_health_status(self):
+        return self.db.get_strict("most_voted_subgraph_health_status")
 
-class CheckSubgraphsHealthRound(CollectionRound):
+    @property
+    def most_voted_subgraph_data(self):
+        return self.db.get_strict("most_voted_subgraph_data")
+
+
+class CheckSubgraphsHealthRound(CollectSameUntilThresholdRound):
     """CheckSubgraphsHealthRound"""
 
     payload_class = CheckSubgraphsHealthPayload
-    payload_attribute = ""  # TODO: update
     synchronized_data_class = SynchronizedData
+    done_event = Event.SYNCHRONIZED
+    collection_key = "subgraph_health_status"
+    selection_key = get_name(SynchronizedData.most_voted_subgraph_health_status)
+
+    _MAX_RETRIES: int = 3
+    retries: int = _MAX_RETRIES
+
+    def reset_retries(self):
+        self.retries = self._MAX_RETRIES
 
     def end_block(self) -> Optional[Tuple[BaseSynchronizedData, Enum]]:
         """Process the end of the block."""
-        synchronized_data = self.synchronized_data
-        return synchronized_data, Event.SYNCHRONIZED
 
+        if self.threshold_reached and self.most_voted_payload is not None:
+            if self.retries == 0:
+                self.reset_retries()
+                return self.synchronized_data, Event.MAX_RETRIES
+            if not self.most_voted_payload:
+                self.retries -= 1
+                return self.synchronized_data, Event.RETRY
```

Comment:
this should also require a reset! basically before ANY return it hit we should reset the class state. And this is why I hate class attributes so much. Nasty nasty. Lets use functional programming wherever we can :pray:

Reply:
state is so nasty. Any state modification (because yes, state is a necessary thing to manage) should be encapsulated and isolated and be performed by a singular process. Jujst image now. We hit zero, we hit -1 on next iteration, we will always hit `DONE` after the fact via ` return super().end_block()` (where we do `reset_retries()`). You see my point? We don't get in a death spiral even, we only log a "proper" happy path cycle next time (using old data!) before recovering. So we might not ever catch this apart from now noticing before the fact. This stuff nauseates me so bad!